Goods update

jPeachPBA opened the pull request:
The main changes are:

- Now requires TELMoS goods files that contain 803 x 803 zones. No longer uses Road Traffic Forecast for external movements.

- Increased the output precision of the add-in DAT files. This allows the Cube smoothing process to recognise if the trip ends are identical.

- Fixed an issue where the precision of the pivot files (3 decimal places) was not the same as the calculated values for the forecast years, these were double precision. Now uses 3 decimal places when calculating the growth in telmos-main

- Split the NRTF.DAT factor file into RTF.DAT (Road Traffic Forecast) and PT.DAT (Public Transport - from Scotland GVA and PDFH elasticities)

- Added GUI options to select RTF.DAT, PT.DAT, and airport_factors.csv. Although they should still default to the existing versions.

- Fixed the outstanding issues: Disable inputs when running, removed rebasing option in GUI, added export/import settings files, added a better subheading with a link to the GitHub releases page.

jPeach-Stantec committed:
Added renumbering to goods file to support new format that includes external zones

jPeach-Stantec committed:
All input widgets now disable when running

jPeach-Stantec committed:
Removed "Rebasing Run" option from GUI. Must now be changed in gui.py.

jPeach-Stantec committed:
Added import and export of settings files

jPeach-Stantec committed:
Split NRTF factors file to RTF.dat (Road Traffic Forecast) for car and PTF (Public Transport Forecast) for PT - PTF.dat values will need updating

jPeach-Stantec committed:
Changed references to NRTF.DAT in telmos_addins to look for separate files for Car and PT growth - RTF.DAT and PTF.DAT
Removed use of NRTF.DAT in telmos_goods - uses TELMoS data only

jPeach-Stantec committed:
Updated GUI with new title string and link to GitHub releases
Added an indeterminate progress bar to show if the process is running

jPeach-Stantec committed:
Changes to GUI.py tkinter structure.
Having the tkinter root object as an attribute of the Application class can cause problems with tkinter objects not being properly tidied.

jPeach-Stantec committed:
Changed addin outputs to be to 9 decimal places
Changed addins file to allow user defined rtf/pt forecast data

jPeach-Stantec committed:
Added user defined rtf and pt forecast data

jPeach-Stantec committed:
Improved issues with numerical precision - Calculating pivot
files on the fly meant they had higher precision than the base versions, these are now rounded when sent to the calculate_growth function

Attractions for the CTE files are now calculated using the base CTE files - previous versions used the TOD files that had lower precision

jPeach-Stantec committed:
Added user options for choosing non-default RTF.DAT and PTF.DAT files

jPeach-Stantec committed:
Disabled the addin DAT files output - now only outputs trip end files instead
These files are not used in the Demand model etc. and speeds up the process

jPeach-Stantec committed:
Added user defined airport growth file
GUI changes

jPeach-Stantec committed:
Removed unnecessary TemPro column from area correspondence csv

jPeach-Stantec committed:
Updated default PT forecast values with values based on Scotland GVA and elasticity from PDFH
Renamed PSV column to PT

asongtoruin commented on gui.py:
```diff
@@ -144,36 +215,109 @@ def callback_run_script(self):
+            self.ptf_file.set(factor_files["PTF File"])
+            self.airport_growth.set(factor_files["Airport Growth File"])
+        except:
+            pass
```
Couple of things here:

1. Can we use json/yml as a format instead of csv?
2. Auto-creating a new directory seems prone to issues, either through permissions or unwitting overwriting of previously-saved settings. For a more direct analogue with loading settings, can we not use `asksaveasfilename`?

asongtoruin replied:
Adding a third thing - we should avoid blanket `except` statements. My guess is that you're trying to avoid a `KeyError` here when the files are not defined in the settings file - in which case that is the exception we should be catching.

jPeachPBA replied:
I've amended this so that it uses `asksaveasfilename` and a json file instead. The rtf/pt/airport factor files are also now in the same format as the other parameters so it is a bit tidier.

asongtoruin commented on gui.py:
```diff
@@ -6,43 +6,69 @@
+    except:
+        pass
+    for child_widget in base.winfo_children():
+        toggle_widgets(child_widget, target_state)
```
Shouldn't use a blanket `except` statement.

jPeachPBA replied:
Agreed, this was meant to catch tkinter `TclError` for when an object doesn't have a state option. Updated in the latest commit.

asongtoruin commented on gui.py:
```diff
@@ -54,9 +80,20 @@ def init_widgets(self):
+        "This tool uses TELMoS planning data and trip rates from NTEM to apply "
+        "growth to the calibrated forecast trip ends that are input to the demand model "
+        "of the Transport Model for Scotland (TMfS)."
+        "The trip end files produced by this tool should be run through the"
```
Spaces missing at the end of these lines, means the text reads oddly.

asongtoruin commented on gui.py:
```diff
@@ -54,9 +80,20 @@ def init_widgets(self):
+                                "tmfs18-trip-end-model/releases")
+        github_link = ttk.Button(title_frame, text="GitHub Releases", 
+                                 command=lambda : open_new(
+                                         github_releases_link))
```
Don't like this as a name for the button. Maybe "Check for newer releases" would be more appropriate, as that is what the user will be trying to do. Adding "Github" into the equation might scare people off from checking.

asongtoruin commented on gui.py:
```diff
@@ -87,53 +124,87 @@ def init_widgets(self):
+        LabelledEntry(base_frame, self.user_names["base_scenario"].split()[1],
+                      self.vars["base_scenario"], 
                       lw=10, w=10, anchor="center", 
                       tool_tip_text="Base scenario name")
```
Why are we using `.split()` here? The "user-friendly names" are only used for this purpose, so we should set them to be the text that we actually want to show. I can't see a reason why having duplicated text in that list would matter at all.

Additionally, this is the same piece of code three times, with the only exception being the tooltips. An argument to be made that we could lose the tooltips and use a loop to generate all three instead?

jPeachPBA replied:
This part and other bits in gui.py are now in a loops to generate the widgets. I've also renamed some of the variables so they don't need to be split.

asongtoruin commented on gui.py:
```diff
@@ -87,53 +124,87 @@ def init_widgets(self):
+        LabelledEntry(tel_frame, self.user_names["tel_scenario"].split()[1],
+                      self.vars["tel_scenario"], 
                       lw=10, w=10, anchor="center", 
                       tool_tip_text="Future scenario name")
```
Same as with base

jPeachPBA replied:
Combined with the base version.

asongtoruin commented on gui.py:
```diff
@@ -87,53 +124,87 @@ def init_widgets(self):
+                                  w=30, lw=25, text_style="HEAD.TLabel")
+        rtf_dir.add_browse(os.getcwd())
+        ptf_dir.add_browse(os.getcwd())
+        air_dir.add_browse(os.getcwd())
```
Again, could be done in a loop

jPeachPBA replied:
Merged this with lines above that creates the widgets - all in a loop now

asongtoruin commented on gui.py:
```diff
@@ -87,53 +124,87 @@ def init_widgets(self):
+        self.log = TextLog(log_frame, width=50, height=25)
+        self.progress = ttk.Progressbar(log_frame, length=280, 
+                                        mode="indeterminate")
+        self.progress.stop()
```
Is this line necessary? Doesn't seem to auto-start.

jPeachPBA replied:
No, I was trying to keep the progress bar empty when initially launched but it cannot be done in indeterminate mode.

asongtoruin commented on gui.py:
```diff
@@ -144,36 +215,109 @@ def callback_run_script(self):
+        self.rtf_file.set(os.path.join(text, "Factors", "RTF.DAT"))
+        self.ptf_file.set(os.path.join(text, "Factors", "PTF.DAT"))
+        self.airport_growth.set(os.path.join(text, "Factors", 
+                                             "airport_factors.csv"))
```
I don't see this being necessary. At present, the functions that use these files use these exact default paths if no link is provided, and we don't provide an interface for users to change these paths.

jPeachPBA replied:
Agreed, I've changed the way that the functions handle the paths to use the default if a blank path is supplied.

asongtoruin commented on telmos_addins.py:
```diff
@@ -29,9 +30,16 @@ def telmos_addins(delta_root, tmfs_root,
+        ptf_file = os.path.join(tmfs_root, "Factors", "PTF.DAT")
+    for factor_file in [rtf_file, ptf_file]:
+        assert os.path.exists(factor_file), \
+                "Addin factor file not found {}".format(factor_file)
```
`assert` shouldn't be used for this. We should be raising an error - in this case, a `FileNotFoundError` - that explains the issue.

asongtoruin replied:
```suggestion
        if not os.path.isfile(factor_file):
            raise FileNotFoundError("File doesn't exist: " + factor_file)
```

jPeachPBA replied:
Ok, updated with this suggestion.

asongtoruin commented on telmos_goods.py:
```diff
@@ -31,15 +31,33 @@ def load_goods_data(goods_file, hgv_output, lgv_output):
     lgv_df = pd.DataFrame(lgv_lines, columns=["I", "J", "V"], dtype="float64")
     
+    assert len(hgv_df) == 803*803, "HGV file requires 803 * 803 entries"
+    assert len(lgv_df) == 803*803, "LGV file requires 803 * 803 entries"
```
Again - this isn't how `assert` should be used.

jPeachPBA replied:
Ok, it raises an `Exception` now. Not sure if there is a better option?

asongtoruin commented on telmos_goods.py:
```diff
@@ -31,15 +31,33 @@ def load_goods_data(goods_file, hgv_output, lgv_output):
+    internal_renumber = {x:(x-16) for x in range(800, 804)}
+    # External zones 784-799 renumbered to 788-803
+    external_renumber = {x:(x+4) for x in range(784, 800)}
+    zone_map = {**zone_map, **internal_renumber, **external_renumber}
```
This could be compressed to a single dict comprehension
```suggestion
    # Internal zones 800-803 get renumbered to 784-787, 
    # and external zones 784-799 renumbered to 788-803
    zone_map = {
        z: z if z < 784 else z+4 if z < 800 else z-16 for z in range(1, 804)
    }
```

asongtoruin commented on telmos_goods.py:
```diff
@@ -31,15 +31,33 @@ def load_goods_data(goods_file, hgv_output, lgv_output):
         df["J"] = df["J"].astype(int)
+        # Renumber zones
+        df["I"] = df["I"].map(zone_map)
+        df["J"] = df["J"].map(zone_map)
```
We can compress this into a one-liner and use a loop:

```suggestion
    # Rewrap zone numbers as integers and renumber zones
    for df in (hgv_df, lgv_df):
        for col in ("I", "J"):
            df[col] = df[col].astype(int).map(zone_map)
```

asongtoruin commented on telmos_goods.py:
```diff
@@ -159,22 +172,14 @@ def telmos_goods(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
-                    ((goods_total["%s_base" % f_key] * goods_totals["TEL_%s" % goods_type]) / 
-                     (goods_total["%s_forecast" % f_key] * goods_totals["BASE_%s" % goods_type]))
-                    )
+            # Road Traaffic Forecast no longer used for external zones - 
```
Spelling

```suggestion
            # Road Traffic Forecast no longer used for external zones - 
```

asongtoruin commented on telmos_main.py:
```diff
@@ -210,11 +210,13 @@ def apply_pivot_files(tod_data, cte_data,
         # These columns do the following
         if j in [0,3,4,7,8]:
-            sw_cte_array[j,:,8] = sw_array[j,:,5]
+            sw_cte_array[j,:,8] = (cte_data[j,:,8] * 
```
Not entirely sure why this has changed from 5 to 8, can you explain?

jPeachPBA replied:
Previously it was using the data from the base TOD file and applying growth to calculate the forecast CTE file. Although as the TOD files are to a lower precision - 3 decimal places instead of 5 - it was not always accurately recreating the base if the same scenario planning data was used.
This now uses the base CTE file to calculate the forecast version. Not sure why this was not done before.

asongtoruin commented on telmos_main.py:
```diff
@@ -403,13 +405,15 @@ def telmos_main(delta_root, tmfs_root, tel_year, tel_id, tel_scenario,
-                                           "airport_factors.csv"), index_col="Year")
+        if airport_growth_file is None:
+            airport_growth_file = os.path.join(tmfs_root, "Factors", 
+                                           "airport_factors.csv")
```
Should we be adding an `os.path.isfile` check here too?

jPeachPBA replied:
Yep, updated.

asongtoruin reviewed (changes requested):
I still need to do some test runs, but these are general comments  and changes for the coding.

jPeach-Stantec committed:
Added factor file variables to var to be more consistent
Updated Github link name
Created most widgets in a loop rather the individually
Export format for settings is now a json file
(Changes not in gui.py are to enable default paths for factor files)

jPeach-Stantec committed:
Updated exception for toggle_widgets and fixed a typo

jPeach-Stantec committed:
Changed assert statements to raise exceptions

jPeach-Stantec committed:
Added a check for the user airport growth file

jPeachPBA committed:
Apply suggestions from code review

Build the renumber zone_map as a single dict comprehension.
Condense code to remap goods zones
Typo corrections

Co-authored-by: Adam Ruszkowski <adam.ruszkowski@gmail.com>

asongtoruin commented on gui.py:
```diff
@@ -284,34 +257,20 @@ def import_settings(self):
         """
         file_path = filedialog.askopenfilename(
                 parent=self.main_frame, title="Select Settings File",
```
```suggestion
            parent=self.main_frame, title="Select Settings File",
```
Reduce indent for params (in conjunction with the next change)

asongtoruin commented on gui.py:
```diff
@@ -284,34 +257,20 @@ def import_settings(self):
         file_path = filedialog.askopenfilename(
                 parent=self.main_frame, title="Select Settings File",
-                filetypes=[("text files", "*.txt")])
+                filetypes=[("json files", "*.json")])
```
```suggestion
            filetypes=[("JSON files (*.json)", "*.json"), ("All files", "*.*")]
```

Make file search options more consistent with Windows standards - having an "all files" option is always useful, just in case.

asongtoruin commented on gui.py:
```diff
@@ -256,23 +230,22 @@ def export_settings(self):
+                                                 title="Save Settings File",
+                                                 defaultextension=".json",
+                                                 filetypes=[("json files", "*.json")],
+                                                 initialdir=output_dir)
```
```suggestion
        file_path = filedialog.asksaveasfilename(
            parent=self.main_frame, title="Save Settings File",
            defaultextension=".json", initialdir=output_dir,
            filetypes=[("JSON files (*.json)", "*.json"), ("All files", "*.*")]
        )     
```

Make file search options more consistent with Windows standards - having an "all files" option is always useful, just in case.

asongtoruin commented on telmos_goods.py:
```diff
@@ -35,26 +35,24 @@ def load_goods_data(goods_file, hgv_output, lgv_output):
+    if not len(hgv_df) == 803*803:
+        raise Exception("HGV file requires 803 * 803 entries")
+    if not len(lgv_df) == 803*803:
+        raise Exception("LGV file requires 803 * 803 entries")
```
```suggestion
    if not len(hgv_df) == 803*803:
        raise ValueError("HGV file requires 803 * 803 entries")
    if not len(lgv_df) == 803*803:
        raise ValueError("LGV file requires 803 * 803 entries")
```

Generic `Exception` calls aren't ideal - I'd view this as an error with the values being passed through, thus `ValueError`?

asongtoruin commented on gui.py:
```diff
@@ -34,22 +37,23 @@ class Application:
                                "Forecast Year", "Forecast ID", "Forecast Scenario",
-                               "Base Year", "Base ID", "Base Scenario"]
+                               "Base Year", "Base ID", "Base Scenario",
+                               "RTF File", "PTF File", "Airport Growth File"]
```
Are these files optional? If so, can we say that in their descriptions?

jPeachPBA committed:
Update telmos_goods.py

Change generic Exception to ValueError

Co-authored-by: Adam Ruszkowski <adam.ruszkowski@gmail.com>

jPeachPBA committed:
Apply suggestions from code review

Improved options for saving settings files

Co-authored-by: Adam Ruszkowski <adam.ruszkowski@gmail.com>

jPeach-Stantec committed:
Added "Optional" text to the factor files frame

asongtoruin commented:
The GUI still seems to load fine for me, so I think I'm comfortable merging it into main and then going on to test the outputs.
